### PR TITLE
DB-11427 SSDS Spark3.0

### PR DESCRIPTION
**Pull request by jpanko1**
# Short Description
Upgrade SSDS for Spark 3.0

# Long Description
Created a spark3.0 folder and basically copied the code from the spark2.4 folder into the new folder.  The main differences are in the build scripts, along with minor differences in the test code and the README.  Made sure the code built, and tested as described below.

# How to test
I went through the steps given in https://github.com/splicemachine/splice-machine-spark-connector/blob/spark3.0/README.md

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -32,7 +32,7 @@ $ sbt clean package
 [success] Total time: 10 s, completed Feb 21, 2020 12:59:56 PM
 ```
 
-You should have the connector available as `target/scala-2.11/splice-machine-spark-connector_2.11-0.3.0-SNAPSHOT.jar`.
+You should have the connector available as `target/scala-2.11/splice-machine-spark-connector_2.11-0.3.0-SNAPSHOT.jar` for the Spark 2.x versions.  For Spark 3.x, it will be `target/scala-2.12/splice-machine-spark-connector_2.12-0.3.0-SNAPSHOT.jar`.
```
Comment (jaceklaskowski) [inline README.md:35]: nit: Start the sentence `For Spark 3.x` on a new line and markdown will make it look pretty. There are two spaces atm.

File: README.md [hunk 2]
```diff
@@ -63,7 +63,7 @@ OpenJDK 64-Bit Server VM (AdoptOpenJDK)(build 25.222-b10, mixed mode)
 
 Start Splice Machine first, e.g. `./start-splice-cluster -p cdh6.3.0 -bl`.
 Remove `-bl` options unless you are starting the Splice Machine instance for the very first time.
-For Spark 2.4, use cdh6.3.0 as the environment.  For Spark 2.3, use hdp3.1.0 .
+For Spark 3.0, use dbaas3.0 as the environment.  For Spark 2.4, use cdh6.3.0.  For Spark 2.3, use hdp3.1.0 .
```
Comment (jaceklaskowski) [inline README.md:66]: A table with all these version would make comprehension easier, wouldn't it?

File: README.md [hunk 3]
```diff
@@ -129,10 +129,11 @@ After you're done with tests, you can stop Splice Machine using `./start-splice-
 
 ## spark-shell
 
-**NOTE**: [spark-2.4.5-bin-hadoop2.7.tgz](https://archive.apache.org/dist/spark/spark-2.4.5/spark-2.4.5-bin-hadoop2.7.tgz) and 
+**NOTE**: [spark-3.0.1-bin-hadoop2.7.tgz](https://archive.apache.org/dist/spark/spark-3.0.1/spark-3.0.1-bin-hadoop2.7.tgz),
```
Comment (jaceklaskowski) [inline README.md:132]: Please note [3.0.2](https://spark.apache.org/news/spark-3-0-2-released.html) is the latest version.

File: README.md [hunk 4]
```diff
@@ -151,7 +152,7 @@ Url
 Type --help for more information.
 ```
 
-**TIP**: Make sure to use the proper versions of Apache Spark 2.4 (or 2.3), Scala 2.11, and Java 1.8.0.
+**TIP**: Make sure to use the proper versions of Java 1.8.0, and Apache Spark 3.0 with Scala 2.12 or Apache Spark 2.4 (or 2.3) with Scala 2.11.
```
Comment (jaceklaskowski) [inline README.md:155]: FYI: [Spark 3 works with Java 11](http://spark.apache.org/docs/latest/#downloading) which gives better performance and is a supported Java version.

File: README.md [hunk 5]
```diff
@@ -191,7 +193,7 @@ $ spark-shell \
 val compatibleSparkVersion = "2.4.5"
 assert(
     spark.version == compatibleSparkVersion,
-    s"The connector works just fine with Spark $compatibleSparkVersion")
+    s"The connector doesn't work with Spark $compatibleSparkVersion")
```
Comment (jaceklaskowski) [inline README.md:196]: Good catch, but...it should be `spark.version` instead (not `compatibleSparkVersion`).

File: spark3.0/build.sbt [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+//scalaVersion := "2.11.6"
+
+lazy val scalaMajorVersion = settingKey[String]("")
+scalaMajorVersion := "2.12"
+
+//val spliceVersion = "2.8.0.1945"
+val spliceVersion = "3.2.0.2001-SNAPSHOT"
+
+// https://github.com/sbt/sbt/issues/5046
+ThisBuild / useCoursier := false
```
Comment (jaceklaskowski) [inline spark3.0/build.sbt:15]: I think it's no longer needed with the recent sbt versions.

File: spark3.0/project/build.properties [hunk 7]
```diff
@@ -0,0 +1 @@
+sbt.version = 1.3.8
```
Comment (jaceklaskowski) [inline spark3.0/project/build.properties:1]: Use the latest version of sbt 1.4.7

File: spark3.0/src/main/java/com/splicemachine/sample/KafkaTopicProducer.java [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+     *             args[3] - Number of threads, each thread will send the number
+     *                        of messages given in args[2]
+     * @throws Exception
+     */
+    public static void main(String[] args) throws Exception {
+        KafkaTopicProducer kp = new KafkaTopicProducer();
+        kp.server = args[0];
+        kp.topic = args[1];
+        kp.totalEvents = Long.parseLong(args[2]);
+	kp.run(Long.parseLong(args[3]));
```
Comment (jaceklaskowski) [inline spark3.0/src/main/java/com/splicemachine/sample/KafkaTopicProducer.java:45]: nit: indent

**Review by dgomezferro — COMMENTED** [review]
Is it possible to extract common code out to a separate submodule? I would expect APIs to be pretty similar between Spark 2.3+ and Spark 3.0

File: README.md (hunk 1 above)
Comment (jpanko1) [reply README.md]: Changed in commit https://github.com/splicemachine/splice-machine-spark-connector/pull/31/commits/978249160be30c456b08c84acb8e6161476b9410

File: README.md (hunk 2 above)
Comment (jpanko1) [reply README.md]: Changed in commit https://github.com/splicemachine/splice-machine-spark-connector/pull/31/commits/978249160be30c456b08c84acb8e6161476b9410

File: README.md (hunk 5 above)
Comment (jpanko1) [reply README.md]: Changed in commit https://github.com/splicemachine/splice-machine-spark-connector/pull/31/commits/978249160be30c456b08c84acb8e6161476b9410

File: README.md (hunk 3 above)
Comment (jpanko1) [reply README.md:140]: Thanks for the info, will plan to upgrade when the splice engine project upgrades.

File: README.md (hunk 4 above)
Comment (jpanko1) [reply README.md:163]: Thanks for the info, will be worth looking into.

**Review by bklo94 — APPROVED** [review]

File: spark3.0/src/main/java/com/splicemachine/sample/KafkaTopicProducer.java (hunk 8 above)
Comment (jpanko1) [reply spark3.0/src/main/java/com/splicemachine/sample/KafkaTopicProducer.java]: Changed in commit https://github.com/splicemachine/splice-machine-spark-connector/pull/31/commits/79eb591fe611c3e080cf221d27d5edb015b759ec

File: spark3.0/project/build.properties (hunk 7 above)
Comment (jpanko1) [reply spark3.0/project/build.properties]: Upgraded in commit https://github.com/splicemachine/splice-machine-spark-connector/pull/31/commits/d39bb4f6fdad8aa5baa8f50d05ab3ff7b248d865

**Comment by jpanko1** [comment]
> Is it possible to extract common code out to a separate submodule? I would expect APIs to be pretty similar between Spark 2.3+ and Spark 3.0

Entered Jira to consolidate common code: https://splicemachine.atlassian.net/browse/DB-11589

**Comment by martinrupp** [comment]
> > Is it possible to extract common code out to a separate submodule? I would expect APIs to be pretty similar between Spark 2.3+ and Spark 3.0
> 
> Entered Jira to consolidate common code: https://splicemachine.atlassian.net/browse/DB-11589

i would also suggest that. i just did a `diff -r spark2.4/src spark3.0/src`, and it's basically just these two changes:
```
diff -r spark2.4/src/test/scala/SpliceSpec.scala spark3.0/src/test/scala/SpliceSpec.scala
60a61,65
>     SparkSession
>       .builder()
>       .master("local[*]")
>       .config("spark.ui.enabled", false)
>       .getOrCreate
```
and this `_root_` one (?)
```
diff -r spark2.4/src/test/scala/splice/SpliceDataSourceStreamingSpec.scala spark3.0/src/test/scala/splice/SpliceDataSourceStreamingSpec.scala
5c5
< import com.splicemachine.spark2.splicemachine.SplicemachineContext
---
> import _root_.com.splicemachine.spark2.splicemachine.SplicemachineContext
```

**Review by martinrupp — APPROVED** [review]

**Review by dgomezferro — APPROVED** [review]

**Review by carolp-503 — APPROVED** [review]
+1

**Pull request merged**